### PR TITLE
Construct V3Number with correct type instead of changing it manually.

### PR DESCRIPTION
**Pull request by mglb**
Requested in https://github.com/verilator/verilator/pull/3521#discussion_r933080244

### CODE REVIEW

File: src/V3Param.cpp [hunk 1]
```diff
@@ -185,19 +185,21 @@ class ParameterizedHierBlocks final {
...
-        // hierOptParamp width.
-        V3Number varNum(pinValuep, hierOptParamp->num().width());
         if (hierOptParamp->isDouble()) {
-            varNum.isDouble(true);
+            double var;
             if (pinValuep->isDouble()) {
-                varNum.opAssign(pinValuep->num());
+                var = pinValuep->num().toDouble();
             } else {  // Cast from integer to real
+                V3Number varNum(pinValuep, 0.0);
```
Comment (wsnyder) [inline src/V3Param.cpp:193]:
```suggestion
                V3Number varNum{pinValuep, 0.0};
```

File: src/V3Param.cpp [hunk 2]
```diff
@@ -185,19 +185,21 @@ class ParameterizedHierBlocks final {
...
                 varNum.opIToRD(pinValuep->num());
+                var = varNum.toDouble();
             }
-            return v3EpsilonEqual(varNum.toDouble(), hierOptParamp->num().toDouble());
+            return v3EpsilonEqual(var, hierOptParamp->num().toDouble());
         } else {  // Now integer type is assumed
+            // Bitwidth of hierOptParamp is accurate because V3Width already caluclated in the
+            // previous run. Bitwidth of pinValuep is before width analysis, so pinValuep is casted
+            // to hierOptParamp width.
+            V3Number varNum(pinValuep, hierOptParamp->num().width());
```
Comment (wsnyder) [inline src/V3Param.cpp:202]:
```suggestion
            V3Number varNum{pinValuep, hierOptParamp->num().width()};
```

**Review by wsnyder — COMMENTED** [review]
Totally reasonable, let's just update constructor style on updated code.

File: src/V3Param.cpp (hunk 1 above)
Comment (mglb) [reply src/V3Param.cpp]: Done

File: src/V3Param.cpp (hunk 2 above)
Comment (mglb) [reply src/V3Param.cpp]: Done

**Pull request merged**